### DISCUSSION TITLE
TS2322: Type 'QueryClient' is not assignable to type 'undefined'.

### DISCUSSION THREAD

**Discussion by ghost** [TanStack/query#2337] [Q&A]
i used nextjs + typescript

i clone the code in my project  https://github.com/tannerlinsley/react-query/blob/master/examples/nextjs/pages/_app.js

then i got an error

![image](https://user-images.githubusercontent.com/6580897/120427293-3e2b2580-c3a4-11eb-90cd-a7ae1e7eaaa2.png)

![image](https://user-images.githubusercontent.com/6580897/120427331-4a16e780-c3a4-11eb-98fd-fa4448ac6212.png)

**Comment by TkDodo** [comment]
you need to pass a generic type to `useRef`. This is not specific to react-query, that's just how react refs work in typescript:

https://stackoverflow.com/questions/60554808/react-useref-with-typescript-and-functional-component

pro tip: you can use `state` over `refs` for one-time initializations. I've blogged about that here: https://tkdodo.eu/blog/use-state-for-one-time-initializations

and also opened a PR for react-query to change the examples. It will work better with typescript, too:

https://github.com/tannerlinsley/react-query/pull/2253/files#diff-f8a7d7140c7558298858c319f0ab8931140c6dd6dbc87598ec4bddc346ad60adL62-R62

**Reply by ghost** [reply]
thank you very much
